File: in_Java/0015 3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Set<List<Integer>> hs = new HashSet<>();
        int t;
        Map<Integer,Integer> hm;
        List<Integer> temp;
        for(int i = 0; i < nums.length; i++){
            t = -nums[i];
            hm = new HashMap<>();
            for(int j = 0; j < nums.length; j++){
                if(j == i) continue;
                if(!hm.containsKey(t - nums[j])) hm.put(nums[j], j);
                else{
                    temp = new ArrayList<>();
                    temp.add(-t);
                    temp.add(t - nums[j]);
                    temp.add(nums[j]);
                    Collections.sort(temp);
                    hs.add(temp);
                }
            }
        }
        for(List<Integer> i : hs) res.add(i);
        return res;
    }
}

File: in_Python/0015 3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        nums.sort()
        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            s,e = i+1, len(nums)-1
            t = -nums[i]
            while s < e:
                if nums[s] + nums[e] == t:
                    res.append([nums[i],nums[s],nums[e]])
                    s += 1
                    while s < e and nums[s] == nums[s-1]:
                        s += 1
                elif nums[s] + nums[e] < t:
                    s += 1
                else:
                    e -= 1
        return res
    